### PR TITLE
Add notebook for OpenAI integration with Elasticsearch

### PR DESCRIPTION
**Pull request by saarikabhasi**
Adds a notebook explaining on downloading open ai dataset, indexing and performing KNN search. 

- [x] Download the wikipedia openai dataset and extract into a dataframe 
- [x] Setup the index + mappings for elasticsearch and index the dataset
- [x] Encode a question with openai ada-02 model and perform a knn search
- [x]  display the results in a pretty way

### CODE REVIEW

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+   "source": [
+    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/saarikabhasi/elasticsearch-labs/blob/notebook/openai-embedding-example/colab-notebooks-examples/integrations/openai/intro.ipynb)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "349e0e74",
+   "metadata": {},
+   "source": [
+    "This notebook illustrates integrating OpenAI with elasticsearch using the official [Elasticsearch Python client](https://www.elastic.co/guide/en/elasticsearch/client/python-api/current/index.html) and perform [kNN search](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html) on OpenAI Wikipedia vector dataset."
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:24]:
```suggestion
    "This notebook demonstrates how to:
    - index the OpenAI Wikipedia vector dataset into Elasticsearch
    - embed a question with the OpenAI [`embeddings`](https://platform.openai.com/docs/api-reference/embeddings) endpoint
    - perform [kNN search](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html) on the Elasticsearch index using the encoded question."
```

Just being a bit clearer about the steps, emphasizing OpenAI components

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+    "* Go to the [Create deployment](https://cloud.elastic.co/deployments/create) page\n",
+    "  * Select Create deployment"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "de32a789",
+   "metadata": {},
+   "source": [
+    "# Connect to Elasticsearch cloud deployment with python client\n",
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:44]:
```suggestion
    "# Connect to Elasticsearch\n",
```

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    "!python3 -m pip install pandas --quiet\n",
+    "!python3 -m pip install wget --quiet"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "80b55952",
+   "metadata": {},
+   "source": [
+    "Let's download OpenAI wikipedia vector database and extract to dataframe"
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:128]: The downloading and extracting step could be under its own Markdown heading?

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+   "source": [
+    "client.indices.create(index=\"wikipedia_vector_index\", mappings=index_mapping)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c2fb582e",
+   "metadata": {},
+   "source": [
+    "# Index data to elasticsearch index\n",
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:222]:
```suggestion
    "# Index data to Elasticsearch index\n",
```

Make sure any other instances are capitalized :)

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    "We will use [Elasticsearch Python Bulk helpers](https://www.elastic.co/guide/en/elasticsearch/client/python-api/current/client-helpers.html) to index data to elasticsearch index. \n",
+    "\n",
+    "We will first convert Pandas dataframe to helpers bulk api iteratable actions.\n"
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:225]: Could be clearer

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+    "    }\n",
+    "}))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "992b6804",
+   "metadata": {},
+   "source": [
+    "# Perform KNN search on dataset"
```
Comment (leemthompo) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:329]: Consider separating the question encoding and kNN search steps into distinct markdown headings :)

**Review by leemthompo — COMMENTED** [review]
Just a few suggestions for clarity and consistency here. I'm not sure if we're actually gonna merge this into Elasticsearch-labs, but a PR is best place to iterate on the content. I'll let @joemcelroy comment on the code :)

File: colab-notebooks-examples/integrations/openai/intro.ipynb (hunk 1 above)
Comment (leemthompo) [reply colab-notebooks-examples/integrations/openai/intro.ipynb:24]: Make sure this renders correctly in actual Markdown, diffs are a bit messy on Notebooks 😄

File: colab-notebooks-examples/integrations/openai/intro.ipynb (hunk 3 above)
Comment (saarikabhasi) [reply colab-notebooks-examples/integrations/openai/intro.ipynb:128]: Yes, will do 👍

File: colab-notebooks-examples/integrations/openai/intro.ipynb (hunk 6 above)
Comment (saarikabhasi) [reply colab-notebooks-examples/integrations/openai/intro.ipynb]: Added in. a9a3d81a92ada6c8f53c09718937f0cdb5060fcc

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cb38a868",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!python3 -m pip install -q Elasticsearch"
```
Comment (joemcelroy) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:59]: install all dependencies all at once. You can do multiple package installs with one command 

Example:
```
!pip install -qU elasticsearch sentence-transformers==2.2.2
```

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+    "      },\n",
+    "      \"content_vector\": {\n",
+    "          \"type\": \"dense_vector\",\n",
+    "          \"dims\": 1536,\n",
+    "          \"index\": \"true\",\n",
+    "          \"similarity\": \"cosine\"\n",
+    "      },\n",
+    "      \"text\": {\"type\": \"text\"},\n",
+    "      \"title\": {\"type\": \"text\"},\n",
+    "      \"url\": { \"type\": \"text\", \"fields\": {\"keyword\": {\"type\": \"keyword\", \"ignore_above\": 256}}},\n",
```
Comment (joemcelroy) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:203]: make url just a keyword field.

**Comment by saarikabhasi** [comment]
Yeah make sense. Will remove threads altogether and add indexing in batch.

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "80b55952",
+   "metadata": {},
+   "source": [
+    "# Download the dataset and extract to dataframe"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (joemcelroy) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:123]: combine the three code cell_types into one, have the imports right at the top.

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+   "execution_count": null,
+   "id": "ca4c51ab",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "EMBEDDING_MODEL = \"text-embedding-ada-002\"\n",
+    "\n",
+    "question = 'How wide is Atlantic ocean?'\n",
```
Comment (joemcelroy) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:333]: how wide is **the** Atlantic ocean?

File: colab-notebooks-examples/integrations/openai/intro.ipynb [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+    "modelOpenAi = openai.Embedding.create(input=question, model=EMBEDDING_MODEL)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c7e6bf5d",
+   "metadata": {},
+   "source": [
+    "# Perform kNN search"
```
Comment (joemcelroy) [inline colab-notebooks-examples/integrations/openai/intro.ipynb:342]: Title should be more around "Performing Queries"

Comment (joemcelroy) [reply colab-notebooks-examples/integrations/openai/intro.ipynb:342]: plus also mention what you're doing here. Links to documentation for KNN search would be useful too.

Comment (leemthompo) [reply colab-notebooks-examples/integrations/openai/intro.ipynb:342]: Should use `##` for all H2s too

**Pull request closed without merging**